src/game/WorldSocket.cpp

```diff
@@ -258,496 +132,149 @@ int WorldSocket::open(void* a)
-    //g.release();
-
-    if (reactor()->schedule_wakeup
-            (this, ACE_Event_Handler::WRITE_MASK) == -1)
-    {
-        sLog.outError("WorldSocket::schedule_wakeup_output");
-        return -1;
-    }
-
-    return 0;
-}
-
-int WorldSocket::ProcessIncoming(WorldPacket* new_pct)
-{
-    MANGOS_ASSERT(new_pct);
-
-    // manage memory ;)
-    ACE_Auto_Ptr<WorldPacket> aptr(new_pct);
+    // Dump received packet.
+    //sLog.outWorldPacketDump(uint32(get_handle()), new_pct->GetOpcode(), new_pct->GetOpcodeName(), new_pct, true);
 
-    const ACE_UINT16 opcode = new_pct->GetOpcode();
+    WorldPacket *pct = new WorldPacket(opcode, validBytesRemaining);
 
-    if (opcode >= NUM_MSG_TYPES)
+    if (validBytesRemaining)
     {
-        sLog.outError("SESSION: received nonexistent opcode 0x%.4X", opcode);
-        return -1;
+        pct->append(InPeak(), validBytesRemaining);
+        ReadSkip(validBytesRemaining);
     }
 
-    if (closing_)
-        return -1;
-
-    // Dump received packet.
-    sLog.outWorldPacketDump(uint32(get_handle()), new_pct->GetOpcode(), new_pct->GetOpcodeName(), new_pct, true);
-
     try
     {
         switch (opcode)
         {
-            case CMSG_PING:
-                return HandlePing(*new_pct);
             case CMSG_AUTH_SESSION:
-                if (m_Session)
+                if (!!m_session)
                 {
-                    sLog.outError("WorldSocket::ProcessIncoming: Player send CMSG_AUTH_SESSION again");
-                    return -1;
+                    sLog.outError("WorldSocket::ProcessIncomingData: Player send CMSG_AUTH_SESSION again");
+                    return false;
                 }
 
-                return HandleAuthSession(*new_pct);
+                return HandleAuthSession(*pct);
+
+            case CMSG_PING:
+                return HandlePing(*pct);
+
             case CMSG_KEEP_ALIVE:
-                DEBUG_LOG("CMSG_KEEP_ALIVE ,size: " SIZEFMTD " ", new_pct->size());
+                DEBUG_LOG("CMSG_KEEP_ALIVE ,size: " SIZEFMTD " ", pct->size());
+
+                return true;
 
-                return 0;
             default:
             {
-                GUARD_RETURN(m_SessionLock, -1);
-
-                if (m_Session != nullptr)
+                if (!m_session)
                 {
-                    // OK ,give the packet to WorldSession
-                    aptr.release();
-                    // WARNING here we call it with locks held.
-                    // Its possible to cause deadlock if QueuePacket calls back
-                    m_Session->QueuePacket(new_pct);
-                    return 0;
+                    sLog.outError("WorldSocket::ProcessIncomingData: Client not authed opcode = %u", uint32(opcode));
+                    return false;
                 }
-                else
+
+                // if the player has logged out, don't bother queueing the packet.  instead, destroy the session >=]
+                if (m_session->PlayerLogout())
```

Comment:
In general I don't think you can assume you will receive any more data from the client.
You now delete the session in ProcessIncomingData but this does not consider the case where the client server connection is suddenly aborted? Or would ProcessIncomingData still be called?